### PR TITLE
DMP-2725: Transcription and media manual deletion audit

### PR DESCRIPTION
**Pull request by mestebanez**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/DMP-2725

### Change description ###

Add manual deletion audit for media and transcription deletion.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/audit/api/AuditApi.java [hunk 1]
```diff
@@ -3,13 +3,27 @@
...
+    default void record(AuditActivity activity, UserAccountEntity userAccountEntity, String additionalData) {
+        record(activity, userAccountEntity, Optional.empty(), Optional.of(additionalData));
+    }
+
+    default void record(AuditActivity activity, UserAccountEntity userAccountEntity) {
+        record(activity, userAccountEntity, Optional.empty(), Optional.empty());
+    }
+
+    default void record(AuditActivity activity, UserAccountEntity userAccountEntity, CourtCaseEntity courtCase) {
+        record(activity, userAccountEntity, Optional.of(courtCase), Optional.empty());
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/darts/audit/api/AuditApi.java:19]: Should be Optional.ofNullable(...)

File: src/test/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceTest.java [hunk 2]
```diff
@@ -396,15 +405,14 @@ void shouldApproveDeletionWhenDocumentExists() {
...
             verify(transcriptionApproveMarkForDeletionValidator).validate(documentId);
             verify(objectAdminActionRepository).save(objectAdminActionEntityArgumentCaptor.capture());
-
             ObjectAdminActionEntity capturedEntity = objectAdminActionEntityArgumentCaptor.getValue();
+            assertEquals(expectedResponse, actualResponse);
             assertTrue(capturedEntity.isMarkedForManualDeletion(), "Entity should be marked for manual deletion");
             assertEquals(userAccount, capturedEntity.getMarkedForManualDelBy(), "Entity's deletion should be marked by the correct user");
             assertNotNull(capturedEntity.getMarkedForManualDelDateTime(), "Entity's deletion datetime should be set");
-
-            assertEquals(expectedResponse, actualResponse);
```
Comment (Ben-Edwards-cgi) [inline src/test/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceTest.java:407]: Why was this removed?

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java [hunk 3]
```diff
@@ -114,6 +121,13 @@ void postAdminApproveMediaMarkedForDeletionShouldReturnSuccess() throws Exceptio
         assertEquals(superAdminUser.getId(), actionResponse.getMarkedForManualDeletionById());
         assertTrue(actionResponse.getIsMarkedForManualDeletion());
 
+        List<AuditEntity> caseExpiredAuditEntries = dartsDatabase.getAuditRepository()
+            .findAll((Specification<AuditEntity>) (root, query, criteriaBuilder) -> criteriaBuilder.and(
+                criteriaBuilder.equal(root.get(AuditEntity_.additionalData), String.valueOf(adminActionEntity.getId())),
+                criteriaBuilder.equal(root.get(AuditEntity_.auditActivity).get("id"), AuditActivity.MANUAL_DELETION.getId())
+            ));
+
+        assertFalse(caseExpiredAuditEntries.isEmpty());
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java:130]: Should validate the content of the audit entry

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java [hunk 4]
```diff
@@ -111,6 +116,14 @@ void testTranscriptionDocumentHideSuccess() throws Exception {
         TranscriptionDocumentHideResponse transcriptionResponse
             = objectMapper.readValue(mvcResult.getResponse().getContentAsByteArray(), TranscriptionDocumentHideResponse.class);
 
+        List<AuditEntity> caseExpiredAuditEntries = dartsDatabase.getAuditRepository()
+            .findAll((Specification<AuditEntity>) (root, query, criteriaBuilder) -> criteriaBuilder.and(
+                criteriaBuilder.equal(root.get(AuditEntity_.additionalData), String.valueOf(adminActionEntity.getId())),
+                criteriaBuilder.equal(root.get(AuditEntity_.auditActivity).get("id"), AuditActivity.MANUAL_DELETION.getId())
+            ));
+
+        assertFalse(caseExpiredAuditEntries.isEmpty());
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java:125]: Should validate the content of the audit entry and ensuring you are only getting the expected number of them added

File: src/main/java/uk/gov/hmcts/darts/audit/api/AuditApi.java (hunk 1 above)
Comment (mestebanez) [reply src/main/java/uk/gov/hmcts/darts/audit/api/AuditApi.java:19]: Assumed if this API call is made then the value is not null but dosent hurt

File: src/test/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceTest.java (hunk 2 above)
Comment (mestebanez) [reply src/test/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceTest.java:407]: it was moved. Checksyle was complaining

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java (hunk 3 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java:130]: Assumed this was good enough seeing as all we control was the activity and the additional data. The query looks specifically for those values so would return an empty list otherwise.

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java:125]: Assumed this was good enough seeing as all we control was the activity and the additional data. The query looks specifically for those values so would return an empty list otherwise.

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java (hunk 3 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java:130]: Is this approach not consistent with the approach taken https://github.com/hmcts/darts-api/pull/2053

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java:125]: Is this approach not consistent with the approach taken https://github.com/hmcts/darts-api/pull/2053

Comment (Ben-Edwards-cgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java:125]: I think we should validate the content of the audits just to make sure we have returned the correct audit entries

**Review by jackmaloney — APPROVED** [review]

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java:128]: Added additional assertions

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java (hunk 3 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java:141]: Added additional assertions

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (Ben-Edwards-cgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java]: I do not see the assertions?

Comment (Ben-Edwards-cgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java]: I do not see these assertions?

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java [hunk 5]
```diff
@@ -127,6 +128,16 @@ void postAdminApproveMediaMarkedForDeletionShouldReturnSuccess() throws Exceptio
...
+        assertNotNull(caseExpiredAuditEntries.get(0).getCreatedBy());
+        assertNotNull(caseExpiredAuditEntries.get(0).getLastModifiedBy());
+        assertNotNull(caseExpiredAuditEntries.get(0).getCreatedDateTime());
+        assertNotNull(caseExpiredAuditEntries.get(0).getLastModifiedDateTime());
+        assertNotNull(caseExpiredAuditEntries.get(0).getLastModifiedDateTime());
+        assertEquals(userIdentity.getUserAccount().getId(), caseExpiredAuditEntries.get(0).getUser().getId());
+        assertNull(caseExpiredAuditEntries.get(0).getCourtCase());
+
         assertFalse(caseExpiredAuditEntries.isEmpty());
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java]: Not needed as verified on line 132

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java]: They are there if you look at the files changed view

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java (hunk 5 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java]: Could you check this again. The line numbers do not seem relevant

Comment (Ben-Edwards-cgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java]: Agreed I no longer see the relevant to this comment (Guessing code change to mark it redundant) happy for this to be closed

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (Ben-Edwards-cgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java]: I can see these now unsure why they were not coming up before. 
My only suggestion would be to add a size() check to ensure you only have one item :)

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java [hunk 6]
```diff
@@ -127,6 +128,16 @@ void postAdminApproveMediaMarkedForDeletionShouldReturnSuccess() throws Exceptio
                 criteriaBuilder.equal(root.get(AuditEntity_.auditActivity).get("id"), AuditActivity.MANUAL_DELETION.getId())
             ));
 
+        // assert additional audit data
+        assertFalse(caseExpiredAuditEntries.isEmpty());
```
Comment (Ben-Edwards-cgi) [inline src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java]: May want a .size() check to ensure you get the correct number of results returned

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java (hunk 5 above)
Comment (Ben-Edwards-cgi) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java]: Line 132 and line 141 seem to be doing the same thing only one of these verification points is needed

File: src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java (hunk 4 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/transcriptions/controller/TranscriptionControllerAdminPostAproveDeletionIntTest.java]: Done

File: src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java (hunk 5 above)
Comment (mestebanez) [reply src/integrationTest/java/uk/gov/hmcts/darts/audio/controller/AudioControllerPostAdminApproveMediaMarkedForDeletionIntTest.java]: Done

**Review by Ben-Edwards-cgi — APPROVED** [review]

File: src/main/resources/db/migration/common/V1_403__manual_deletion_audit_activity.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO audit_activity (aua_id, activity_name, activity_description, created_ts, created_by, last_modified_ts, last_modified_by)
+VALUES (36, 'Manual Deletion', 'Manual Deletion', CURRENT_TIMESTAMP, 0, CURRENT_TIMESTAMP, 0);
```
Comment (jackmaloney) [inline src/main/resources/db/migration/common/V1_403__manual_deletion_audit_activity.sql:2]: Duplicate of https://github.com/hmcts/darts-api/pull/2133/files#diff-0599014aa9b7da6978e32927fdd55ef52e3b23eb55ce641dcede54a50db981cc Ensure only merged once.

**Review by davet1985 — APPROVED** [review]
